network/proxy.md

```diff
@@ -4,44 +4,49 @@ description: How to configure the Docker client to use a proxy server
 keywords: network, networking, proxy, client
 ---
 
+> **Note**
+> 
+> This topic describes how to configure a proxy server for your containers. 
+> Refer to the [configure the docker daemon to use a proxy server](../config/daemon/systemd.md#configure-the-docker-daemon-to-use-a-proxy-server)
+> section to learn how to configure a proxy server for the Docker daemon
+> (`dockerd`) itself.
+
 If your container needs to use an HTTP, HTTPS, or FTP proxy server, you can
 configure it in different ways:
 
-- In Docker 17.07 and higher, you can
-  [configure the Docker client](#configure-the-docker-client) to pass
-  proxy information to containers automatically.
-
-- In Docker 17.06 and earlier versions, you must set the appropriate
-  [environment variables](#use-environment-variables)
-  within the container. You can do this when you build the image (which makes
-  the image less portable) or when you create or run the container.
+- [configure the Docker client](#configure-the-docker-client) to pass proxy
+  information to containers automatically.
+- [set appropriate environment variables](#use-environment-variables)
+  within the container.
 
 ## Configure the Docker client
 
 1.  On the Docker client, create or edit the file `~/.docker/config.json` in the
     home directory of the user that starts containers. Add JSON similar to the
-    following example. Substitute the type of proxy with `httpsProxy` or `ftpProxy` if necessary, and substitute the address and port of the proxy server. You can also configure multiple proxy servers simultaneously.
+    following example. Substitute the type of proxy with `httpsProxy` or `ftpProxy`
+    if necessary, and substitute the address and port of the proxy server. You
+    can also configure multiple proxy servers simultaneously.
 
     You can optionally exclude hosts or ranges from going through the proxy
     server by setting a `noProxy` key to one or more comma-separated IP
-    addresses or hosts. Using the `*` character as a wildcard for hosts and using CIDR notation for IP addresses is supported as
-    shown in this example:
+    addresses or hosts. Using the `*` character as a wildcard for hosts, and
+    using CIDR notation for IP addresses is supported as shown in this example:
 
     ```json
     {
-     "proxies":
-     {
-       "default":
-       {
+     "proxies": {
+       "default": {
          "httpProxy": "http://192.168.1.12:3128",
-         "httpsProxy": "http://192.168.1.12:3128",
-         "noProxy": "*.test.example.com,.example2.com,127.0.0.0/8"
+         "httpsProxy": "https://192.168.1.12:3129",
```

Comment:
I think it is supposed to be `http` here, quoting @djs55:

> Confusingly a HTTPS proxy is a HTTP server which accepts CONNECT host:port HTTP requests and establishes a tunnel, which the client then sends TLS back and forth but the proxy doesn't understand (in some ways this is easier than in the HTTP case where the proxy has a bunch of rules about what to do with headers)

Reply:
I think both should work, although it _could_ depend on the implementation.

It's really a "fuzzy" area, without any (good) standards. e.g. Golang also made a silly (and breaking) change and started to ignore `HTTP_PROXY` for https connections; https://docs.docker.com/engine/release-notes/#201010

Reply:
I think
- the `httpProxy` (`HTTP_PROXY`) will be used to proxy http:// connections
- the `httpsProxy` (`HTTPS_PROXY`) will be used to proxy https:// connections

where the connection to the proxy can either be plaintext `http://proxy:3128` or itself using TLS `https://proxy:3128`. So there can be 0, 1 or 2 layers of TLS.

If the underlying connection is TLS, then the only reason to use TLS to talk to the proxy too is to protect the authorization header which could include a base64-encoded password. The destination server is sent in the clear even with TLS (as a Server Name Indication) so the `HTTP CONNECT <server:443>` part is not secret.

So I think the examples are ok. My main problem with the docs as they are is that there are a confusing number of places where proxies can be configured. I think this PR makes an improvement though!